src/Custom/Microsoft.Extensions.AI.OpenAIAdapter.cs

```diff
@@ -0,0 +1,1820 @@
+            {
+                AdditionalProperties = new() { [nameof(audio.ExpiresAt)] = audio.ExpiresAt },
+            };
+
+            if (audio.Id is string id)
+            {
+                dc.AdditionalProperties[nameof(audio.Id)] = id;
+            }
+
+            if (audio.Transcript is string transcript)
+            {
+                dc.AdditionalProperties[nameof(audio.Transcript)] = transcript;
+            }
+
+            returnMessage.Contents.Add(dc);
+        }
+
+        // Also manufacture function calling content items from any tool calls in the response.
+        if (options?.Tools is { Count: > 0 })
+        {
+            foreach (ChatToolCall toolCall in openAICompletion.ToolCalls)
+            {
+                if (!string.IsNullOrWhiteSpace(toolCall.FunctionName))
+                {
+                    var callContent = ParseCallContentFromBinaryData(toolCall.FunctionArguments, toolCall.Id, toolCall.FunctionName);
+                    callContent.RawRepresentation = toolCall;
+
+                    returnMessage.Contents.Add(callContent);
+                }
+            }
+        }
+
+        // Wrap the content in a ChatResponse to return.
+        var response = new ChatResponse(returnMessage)
+        {
+            CreatedAt = openAICompletion.CreatedAt,
+            FinishReason = FromOpenAIFinishReason(openAICompletion.FinishReason),
+            ModelId = openAICompletion.Model,
+            RawRepresentation = openAICompletion,
+            ResponseId = openAICompletion.Id,
+        };
+
+        if (openAICompletion.Usage is ChatTokenUsage tokenUsage)
+        {
+            response.Usage = FromOpenAIUsage(tokenUsage);
+        }
+
+        if (openAICompletion.ContentTokenLogProbabilities is { Count: > 0 } contentTokenLogProbs)
+        {
+            (response.AdditionalProperties ??= [])[nameof(openAICompletion.ContentTokenLogProbabilities)] = contentTokenLogProbs;
+        }
+
+        if (openAICompletion.Refusal is string refusal)
+        {
+            (response.AdditionalProperties ??= [])[nameof(openAICompletion.Refusal)] = refusal;
+        }
+
+        if (openAICompletion.RefusalTokenLogProbabilities is { Count: > 0 } refusalTokenLogProbs)
+        {
+            (response.AdditionalProperties ??= [])[nameof(openAICompletion.RefusalTokenLogProbabilities)] = refusalTokenLogProbs;
+        }
+
+        if (openAICompletion.SystemFingerprint is string systemFingerprint)
+        {
+            (response.AdditionalProperties ??= [])[nameof(openAICompletion.SystemFingerprint)] = systemFingerprint;
+        }
+
+        return response;
+    }
+
+    /// <summary>Converts an extensions options instance to an OpenAI options instance.</summary>
+    private static ChatCompletionOptions ToOpenAIOptions(ChatOptions? options)
+    {
+        ChatCompletionOptions result = new();
+
+        if (options is not null)
+        {
```

Comment:
options,Model is currently ignored for all of these. Now that it's in the same assembly, it should be possible to override the model selection.